### ISSUE TITLE
PublishTestResults says test result file is invalid if there are no results

### ISSUE THREAD

**Issue by tillig** [microsoft/azure-pipelines-tasks#9292]
In our Azure Pipelines build we run .NET Core unit tests with a filter against the whole solution:

`dotnet test OurSolution.sln -c Release --filter Stage=Unit --results-directory $(Build.ArtifactStagingDirectory)/log --logger:trx`

This runs all the xunit tests in all test projects in the whole solution that have a trait `Stage` with the value `Unit`. (Other tests run at other times in the build/deployment pipeline so have other values for `Stage`.)

This can lead to a `.trx` file coming out that doesn't actually have any test results in it. It's a valid file, but has no results.

```xml
<?xml version="1.0" encoding="utf-8"?>
<TestRun id="87241152-e534-4aa1-934d-e3ee181831d0" name="@fv-az629 2019-01-09 19:32:39" xmlns="http://microsoft.com/schemas/VisualStudio/TeamTest/2010">
  <Times creation="2019-01-09T19:32:39.0548207+00:00" queuing="2019-01-09T19:32:39.0548209+00:00" start="2019-01-09T19:32:36.3914830+00:00" finish="2019-01-09T19:32:39.0689729+00:00" />
  <TestSettings name="default" id="fe1e920e-1b69-4698-a873-f5efe6b6e633">
    <Deployment runDeploymentRoot="_fv-az629_2019-01-09_19_32_39" />
  </TestSettings>
  <TestLists>
    <TestList name="Results Not in a List" id="8c84fa94-04c1-424b-9868-57a2d4851a1d" />
    <TestList name="All Loaded Results" id="19431567-8539-422a-85d7-44ee4e166bda" />
  </TestLists>
  <ResultSummary outcome="Completed">
    <Counters total="0" executed="0" passed="0" failed="0" error="0" timeout="0" aborted="0" inconclusive="0" passedButRunAborted="0" notRunnable="0" notExecuted="0" disconnected="0" warning="0" completed="0" inProgress="0" pending="0" />
    <Output>
      <StdOut>[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.1 (64-bit .NET Core 4.6.27110.04)[xUnit.net 00:00:00.65]   Discovering: CalculatorService.Behavior[xUnit.net 00:00:00.74]   Discovered:  CalculatorService.Behavior[xUnit.net 00:00:00.76]   Starting:    CalculatorService.Behavior[xUnit.net 00:00:00.88]   Finished:    CalculatorService.Behavior</StdOut>
    </Output>
    <RunInfos>
      <RunInfo computerName="fv-az629" outcome="Warning" timestamp="2019-01-09T19:32:38.9335445+00:00">
        <Text>No test matches the given testcase filter `Stage=Unit` in /home/vsts/work/1/s/test/CalculatorService.Behavior/bin/Release/netcoreapp2.2/CalculatorService.Behavior.dll</Text>
      </RunInfo>
    </RunInfos>
  </ResultSummary>
</TestRun>
```

Later, when we publish the test results from the pipeline...

```yaml
  - task: PublishTestResults@2
    displayName: 'Publish test results'
    condition: always()
    inputs:
      testResultsFormat: VSTest
      testResultsFiles: $(Build.ArtifactStagingDirectory)/log/*.trx
```

...we get a warning on the build:

`##[warning]Invalid results file. Make sure the result format of the file '/home/vsts/work/1/a/log/_fv-az629_2019-01-09_19_32_39.trx' matches 'VSTest' test results format.`

The file itself is valid, so the message here is misleading. Ideally I wouldn't see a warning here; but if we need a warning, it'd be better if it was a more accurate message like "No test results found in the file" or something similar.

**Comment by puagarwa** [comment]
Hi, warning is as per our design of task since we wants to warn customer if he is trying to use PTR task without any results, but i agree with you about more apt warning message and we will correct it in upcoming release.

**Comment by puagarwa** [comment]
#2222

**Comment by puagarwa** [comment]
We have fixed the issue, please try again and let us know if there is any issue.

**Comment by SadagopanRajaram** [comment]
Closing the issue - please do reopen if this is not resolved

**Comment by Anakha77** [comment]
Still having this misleading message with Visual Studio Test task.
Maybe it is the same issue since this task run tests and publish test results.

**Comment by puagarwa** [comment]
@Anakha77  can you please try with latest agent once. if you still face the issue please provide these information

you are using onprem or VSTS - if onprem then what is the version,
what is the agent version you are using.
if possible please share your debug logs for your vstest.
if you can also give more information about your scenario , it will be great.

**Comment by Sup3rLativ3** [comment]
Version: VSTS/Azure DevOps
Current agent version: '2.150.3'

```
##[section]Starting: pytest
==============================================================================
Task         : Command Line
Description  : Run a command line script using cmd.exe on Windows and bash on macOS and Linux.
Version      : 2.148.0
Author       : Microsoft Corporation
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613735)
==============================================================================
Generating script.
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/ebf778eb-c686-462f-a7e6-872f39e43612.sh
Collecting pytest
  Downloading https://files.pythonhosted.org/packages/56/53/0ae37ab12c457945ae0152c6571d6d40eecccddf25f71fe328f9aefe90ca/pytest-4.5.0-py2.py3-none-any.whl (227kB)
Collecting pytest-azurepipelines
  Downloading https://files.pythonhosted.org/packages/18/cd/5931a8f4fa228086a2dd0459226c72fee79f4877249859f48a1a5266a8fc/pytest_azurepipelines-0.6.0-py3-none-any.whl
Collecting atomicwrites>=1.0 (from pytest)
  Downloading https://files.pythonhosted.org/packages/52/90/6155aa926f43f2b2a22b01be7241be3bfd1ceaf7d0b3267213e8127d41f4/atomicwrites-1.3.0-py2.py3-none-any.whl
Collecting py>=1.5.0 (from pytest)
  Downloading https://files.pythonhosted.org/packages/76/bc/394ad449851729244a97857ee14d7cba61ddb268dce3db538ba2f2ba1f0f/py-1.8.0-py2.py3-none-any.whl (83kB)
Collecting pathlib2>=2.2.0; python_version < "3.6" (from pytest)
  Downloading https://files.pythonhosted.org/packages/2a/46/c696dcf1c7aad917b39b875acdc5451975e3a9b4890dca8329983201c97a/pathlib2-2.3.3-py2.py3-none-any.whl
Requirement already satisfied: six>=1.10.0 in /opt/hostedtoolcache/Python/3.5.5/x64/lib/python3.5/site-packages (from pytest) (1.12.0)
Collecting wcwidth (from pytest)
  Downloading https://files.pythonhosted.org/packages/7e/9f/526a6947247599b084ee5232e4f9190a38f398d7300d866af3ab571a5bfe/wcwidth-0.1.7-py2.py3-none-any.whl
Collecting more-itertools>=4.0.0; python_version > "2.7" (from pytest)
  Downloading https://files.pythonhosted.org/packages/b3/73/64fb5922b745fc1daee8a2880d907d2a70d9c7bb71eea86fcb9445daab5e/more_itertools-7.0.0-py3-none-any.whl (53kB)
Collecting pluggy!=0.10,<1.0,>=0.9 (from pytest)
  Downloading https://files.pythonhosted.org/packages/06/ee/de89e0582276e3551df3110088bf20844de2b0e7df2748406876cc78e021/pluggy-0.12.0-py2.py3-none-any.whl
Collecting attrs>=17.4.0 (from pytest)
  Downloading https://files.pythonhosted.org/packages/23/96/d828354fa2dbdf216eaa7b7de0db692f12c234f7ef888cc14980ef40d1d2/attrs-19.1.0-py2.py3-none-any.whl
Requirement already satisfied: setuptools in /opt/hostedtoolcache/Python/3.5.5/x64/lib/python3.5/site-packages (from pytest) (28.8.0)
Collecting importlib-metadata>=0.12 (from pluggy!=0.10,<1.0,>=0.9->pytest)
  Downloading https://files.pythonhosted.org/packages/63/da/4270fe75ba542de6a350c59339988e4a5275d3efff80b704ca2b5ede8120/importlib_metadata-0.15-py2.py3-none-any.whl
Collecting zipp>=0.5 (from importlib-metadata>=0.12->pluggy!=0.10,<1.0,>=0.9->pytest)
  Downloading https://files.pythonhosted.org/packages/a0/0f/9bf71d438d2e9d5fd0e4569ea4d1a2b6f5a524c234c6d221b494298bb4d1/zipp-0.5.1-py2.py3-none-any.whl
Installing collected packages: atomicwrites, py, pathlib2, wcwidth, more-itertools, zipp, importlib-metadata, pluggy, attrs, pytest, pytest-azurepipelines
Successfully installed atomicwrites-1.3.0 attrs-19.1.0 importlib-metadata-0.15 more-itertools-7.0.0 pathlib2-2.3.3 pluggy-0.12.0 py-1.8.0 pytest-4.5.0 pytest-azurepipelines-0.6.0 wcwidth-0.1.7 zipp-0.5.1
============================= test session starts ==============================
platform linux -- Python 3.5.5, pytest-4.5.0, py-1.8.0, pluggy-0.12.0
rootdir: /home/vsts/work/1/s
plugins: azurepipelines-0.6.0
collected 0 items
Timestamp is not available for one or more testsuites. Total run duration is being calculated as the sum of time durations of detected testsuites
##[warning]Invalid results file. Make sure the result format of the file '/home/vsts/work/1/s/test-output.xml' matches 'JUnit' test results format.

----------- generated xml file: /home/vsts/work/1/s/test-output.xml ------------
========================= no tests ran in 0.06 seconds =========================
##[error]Bash exited with code '5'.
##[section]Async Command Start: Publish test results
##[section]Async Command End: Publish test results
##[section]Finishing: pytest
```

**Comment by puagarwa** [comment]
@Sup3rLativ3 couple of things for above log.
1. This log is not vstest task and it seems to be some custom task.
2. we have fixed this issue in 151 , so you need to use agent version 151 or above. please try and let us know.
3. if the 151 agent doesnt resolve the issue then the problem could also because of result file , in that case please do share your junit test.

**Comment by Anakha77** [comment]
@puagarwa I use Azure DevOps Service and Hosted VS2017 agent.
I ended up with no results in my test execution because of a bad targeted framework for test run

```Starting: Run Tests
==============================================================================
Task         : Visual Studio Test
Description  : Run unit and functional tests (Selenium, Appium, Coded UI test, etc.) using the Visual Studio Test (VsTest) runner. Test frameworks that have a Visual Studio test adapter such as MsTest, xUnit, NUnit, Chutzpah (for JavaScript tests using QUnit, Mocha and Jasmine), etc. can be run. Tests can be distributed on multiple agents using this task (version 2).
Version      : 2.151.6
Author       : Microsoft Corporation
Help         : [More information](https://go.microsoft.com/fwlink/?LinkId=835764)
==============================================================================
SystemVssConnection exists true
SystemVssConnection exists true
SystemVssConnection exists true
Running tests using vstest.console.exe runner.
======================================================
Test selector : Test assemblies
Test filter criteria : null
Search folder : D:\a\1\s
VisualStudio version selected for test execution : latest
Attempting to find vstest.console from a visual studio installation.
Attempting to find vstest.console from a visual studio build tools installation.
Attempting to find vstest.console from a visual studio installation.
Run in parallel : false
Run in isolation : false
Path to custom adapters : null
Other console options : null
Code coverage enabled : true
Diagnostics enabled : true
SystemVssConnection exists true
Run the tests locally using vstest.console.exe
========================================================
Test selector : Test assemblies
Test assemblies : **\*test*.dll,!**\*TestAdapter.dll,!**\obj\**
Test filter criteria : null
Search folder : D:\a\1\s
Run settings file : D:\a\1\s
Run in parallel : false
Run in isolation : false
Path to custom adapters : null
Other console options : null
Code coverage enabled : true
Diagnostics enabled : false
Rerun failed tests: false
VisualStudio version selected for test execution : latest
Attempting to find vstest.console from a visual studio installation.
Attempting to find vstest.console from a visual studio build tools installation.
Attempting to find vstest.console from a visual studio installation.
========================================================
======================================================
:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\CommonExtensions\Microsoft\TestWindow\vstest.console.exe" @D:\a\_temp\bf742f51-7e65-11e9-a374-25918adf5845.txt
Microsoft (R) Test Execution Command Line Tool Version 15.9.1
Copyright (c) Microsoft Corporation.  All rights reserved.

vstest.console.exe 
"D:\a\1\s\DataCollector.Tests\bin\Release\netcoreapp2.1\DataCollector.Tests.dll"
"D:\a\1\s\DataCollector.Tests\bin\Release\netcoreapp2.1\Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.Interface.dll"
"D:\a\1\s\DataCollector.Tests\bin\Release\netcoreapp2.1\Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.dll"
/EnableCodeCoverage
/logger:"trx"
/TestAdapterPath:"D:\a\1\s"
Starting test execution, please wait...
Test run will use DLL(s) built for framework .NETFramework,Version=v4.0 and platform X86. Following DLL(s) do not match framework/platform settings.
DataCollector.Tests.dll is built for Framework 2.1 and Platform AnyCPU.
Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.Interface.dll is built for Framework 4.5 and Platform AnyCPU.
Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.dll is built for Framework 1.5 and Platform AnyCPU.
Go to http://go.microsoft.com/fwlink/?LinkID=236877&clcid=0x409 for more details on managing these settings.

Microsoft (R) Coverage Collection Tool Version 15.0.30319.200


Copyright (c) Microsoft Corporation.  All rights reserved.




2.9604
No test is available in D:\a\1\s\DataCollector.Tests\bin\Release\netcoreapp2.1\DataCollector.Tests.dll D:\a\1\s\DataCollector.Tests\bin\Release\netcoreapp2.1\Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.Interface.dll D:\a\1\s\DataCollector.Tests\bin\Release\netcoreapp2.1\Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
Microsoft (R) Coverage Collection Tool Version 15.0.30319.200


Copyright (c) Microsoft Corporation.  All rights reserved.




Results File: D:\a\1\s\TestResults\VssAdministrator_fv-az467_2019-05-24_20_52_04.trx

Attachments:
  D:\a\1\s\TestResults\7a061b1b-b49e-46cf-97d6-f9ab26055603\VssAdministrator_fv-az467 2019-05-24 20_51_54.coverage

##[warning]Invalid results file. Make sure the result format of the file 'D:\a\1\s\TestResults\VssAdministrator_fv-az467_2019-05-24_20_52_04.trx' matches 'VSTest' test results format.
Async Command Start: Publish test results
Async Command End: Publish test results
Finishing: Run Tests```

**Comment by puagarwa** [comment]
@Anakha77 I looked around and found the hosted agent version is 150.3 and our fix got merged into 151 so wait for few days, and the issue will be fixed once hosted agent get updated, meanwhile if you want to verify the fix, you can use your local agent with 151 or later version.
